06_gpu_and_ml/stable_lm/main.py

```diff
@@ -0,0 +1,322 @@
+import base64
+import os
+import time
+from pathlib import Path
+from typing import Any, Dict, Generator, List, Optional, Union
+
+from pydantic import BaseModel
+from typing_extensions import Annotated, Literal
+
+import modal
+
+with open("requirements.txt", "r") as f:
+    requirements = base64.b64encode(f.read().encode("utf-8")).decode("utf-8")
+
+
+def build_models():
+    import torch
+    from huggingface_hub import snapshot_download
+    from transformers import AutoModelForCausalLM, AutoTokenizer
+
+    model_path = snapshot_download(
+        "stabilityai/stablelm-tuned-alpha-7b",
+        ignore_patterns=["*.md"],
+    )
+    m = AutoModelForCausalLM.from_pretrained(
+        model_path,
+        torch_dtype=torch.float16,
+        device_map="auto",
+        local_files_only=True,
+    )
+    m.save_pretrained(
+        model_path, safe_serialization=True, max_shard_size="24GB"
+    )
+    tok = AutoTokenizer.from_pretrained(model_path)
+    tok.save_pretrained(model_path)
+    [p.unlink() for p in Path(model_path).rglob("*.bin")]  # type: ignore
+
+
+image = (
+    modal.Image.conda()
+    .apt_install("git", "software-properties-common", "wget")
+    .conda_install(
+        "cudatoolkit-dev=11.7",
+        "pytorch-cuda=11.7",
+        "rust=1.69.0",
+        channels=["nvidia", "pytorch", "conda-forge"],
+    )
+    .env(
+        {
+            "HF_HOME": "/root",
+            "HF_HUB_ENABLE_HF_TRANSFER": "1",
```

Comment:
Nice. Did you observe a speed up from this? If so, how much?

Reply:
I'd have to quantify. During build, I didn't notice much of a change for this file size but locally on an M1 I do find it quicker for fetching larger model files.